add grub password procedure (#64558)
openshift/openshift-docs: modules/rhcos-setting-grub-password-day-2.adoc

// Module included in the follow assemblies:
//
// * post_installation_configuration/machine-configuration-tasks.adoc

:_content-type: PROCEDURE
[id="rhcos-setting-grub-password-day-2_{context}"]
= Setting a GRUB password on {op-system}

:FeatureName: GRUB password management
include::snippets/technology-preview.adoc[]

You can set up a password to prevent unauthorized users from accessing the GRUB command line, modify kernel command-line arguments, or boot non-default OSTree deployments.

.Prerequisites

> * You have a running {product-title} cluster.

Reviewer: Does this functionality work on cluster that were older and upgraded to 4.14? If there are any cornoer case, we may want to document that with keeping in mind node-scale-up happen with older bootimage.

Reviewer 2: We will only support that on clusters installed with 4.14 for now.

Author: I can clarify this in the opening paragraph

* You are logged in as a cluster administrator.
* You installed the OpenShift CLI (`oc`).
* You installed the `butane` utility.
* You installed the `grub2-tools-minimal` package.

.Procedure

. Create a password hash for GRUB using the `grub2-mkpasswd-pbkdf2` command line tool:
+
[source,terminal]
----
$ grub2-mkpasswd-pbkdf2
Enter password:
Reenter password:
PBKDF2 hash of your password is
grub.pbkdf2.sha512.10000.5AE6255...
----

. Create a script that validates your configurations and sets the GRUB config for the root user with the newly created password:
+
.rhcos_set_grub_password.sh
[source,bash]
----

> set -euo pipefail

Reviewer 2: ```suggestion
#!/bin/bash
set -euo pipefail
```

STAMPFILE="/var/lib/.grub_password_<date_changed>"
GRUBPASSWORD="grub.pbkdf2.sha512.10000.58E6255..."

main () {

# Validate GRUB config

>   if [[ $(grep -c 'source $prefix/user.cfg' '/boot/grub2/grub.cfg') != 1]]; then

Reviewer 3: ```suggestion
  if [[ $(grep -c 'source $prefix/user.cfg' '/boot/grub2/grub.cfg') != 1 ]]; then
```

Reviewer 3: exec error
```
/usr/local/bin/rhcos_set_grub_password.sh: line 10: syntax error in conditional expression: unexpected token `;'
```

    echo "Invalid GRUB config: Missing user.cfg"
    return 1
  fi

# Validate ostree repo config
  if [[ $(grep -c 'bls-append-except-default=grub_users=""' '/sysroot/ostree/repo/config') != 1 ]]; then

>     echo "Invalid ostree repo config: Missing bls-append-except-default"

Reviewer 4: Mmmm right...we don't add this across upgrades, but we probably should.

    return 1
  fi

# Remount /boot as RW
  mount -o remount,rw /boot

# Write config

>   echo -e "set superusers="root"\npassword_pbkdf2 root {GRUBPASSWORD}" > /boot/grub2/user.cfg

Reviewer 2: ```suggestion
  echo -e "set superusers=\"root\"\npassword_pbkdf2 root ${GRUBPASSWORD}" > /boot/grub2/user.cfg
```

  chmod 400 /boot/grub2/user.cfg

# Write stamp file

>   touch "{STAMPFILE}"

Reviewer 2: ```suggestion
  touch "${STAMPFILE}"
```

> }

Reviewer 2: ```suggestion
}

main "${@}"
```

----

. Create a systemd service unit that launches the script on boot:
+
.rhcos-set-grub-password.service
[source,text]
----
[Unit]
Description=Set GRUB password
ConditionPathExists=!/var/lib/.grub_password_<date_changed>

> RequiresMountsFor=/boot

Reviewer 2: We might need to add:
```
Before=kubelet.service
```

Will confirm that.

Reviewer 2: @sinnykumari Is that enough to order before the MCD does any update operation on first boot? What do we usually use to order units before the kubelet / MCD starts?

Reviewer: If we want it to run before MCD starts any update during firstboot, maybe just add `Before=machine-config-daemon-firstboot.service` ? See, https://github.com/openshift/machine-config-operator/blob/master/templates/common/_base/units/machine-config-daemon-firstboot.service.yaml .
cc @cgwalters WDYT?

Reviewer: umm, but wait isn't setting grub password can be a day1 or day2 operation?

Reviewer 2: Yes, it can be both as once set up day-2 it will impact new nodes on day-1

Reviewer: if it's latter, I think we should add both `Before=machine-config-daemon-firstboot.service kubelet.service` . As per my understanding, here machine-config-daemon-firstboot.service not running during follow-up boots shouldn't be causing this unit to fail.

> RequiresMountsFor=/boot

Reviewer 2: ```suggestion
RequiresMountsFor=/boot
Before=machine-config-daemon-firstboot.service kubelet.service
```

[Service]
Type=oneshot
RemainAfterExit=yes
MountFlags=slave
ExecStart=/usr/local/bin/rhcos_set_grub_password.sh

[Install]
WantedBy=multi-user.target
----

. Create a Butane config that includes the files from the previous steps:
+

> .40-worker-custom-journald.bu

Reviewer 2: ```suggestion
.40-worker-grub-password.bu
```

[source,yaml]
----
variant: openshift

> version: {product-version}.0

Reviewer 3: suggest to use 4.13.0, it will generate machine config with ignition ver. 3.2.0. rendered config version is 3.4.0  [4.14.0-experimental](https://coreos.github.io/butane/config-openshift-v4_14-exp/) it not recommended, because it will generate machine config with ignition ver `3.5.0-experimental` which is not supported in 4.14

Reviewer 2: We'll fix that soon and downgrade 4.14 to 3.4: https://github.com/coreos/butane/issues/488
This should stay the on 4.14.0 variant.

metadata:
  name: 99-worker-grub-password

>   labels: machineconfiguration.openshift.io/role: worker

Reviewer 2: ```suggestion
  labels:
    machineconfiguration.openshift.io/role: worker
```

storage:
  files:
    - path: /usr/local/bin/rhcos_set_grub_password.sh
      mode: 0700
      overwrite: true
      contents:
        inline: |

>         ... <1>

Reviewer 2: ```suggestion
          ... <1>
```

>     - path: /etc/systemd/system/rhcos-set-grub-password.service
>       mode: 0644
>       overwrite: true

Reviewer 2: ```suggestion
systemd:
  units:
    - name: rhcos-set-grub-password.service
      enabled: true
```

      contents:

>         inline: |

Reviewer 2: ```suggestion
```

        ... <2>
----
<1> The contents of the file `rhcos_set_grub_password.sh`.
<2> The contents of the file `rhcos-set-grub-password.service`.

> . Generate a `MachineConfig` object, `40-worker-custom-journald.yaml`, that contains the configuration you want to deliver to the nodes:

Reviewer 2: ```suggestion
. Generate a `MachineConfig` object, `40-worker-grub-password.yaml`, that contains the configuration you want to deliver to the nodes:
```

+
[source,terminal]
----

> $ butane 40-worker-custom-journald.bu -o 40-worker-custom-journald.yaml

Reviewer 2: ```suggestion
$ butane 40-worker-grub-password.bu -o 40-worker-grub-password.yaml
```

----

. Apply the machine config:
+
[source,terminal]
----

> $ oc apply -f 40-worker-custom-journald.yaml

Reviewer 2: ```suggestion
$ oc apply -f 40-worker-grub-password.yaml
```

----
+
The controller detects the new `MachineConfig` object and generated a new `rendered-worker-<hash>` version.

. Monitor the status of the rollout of the new rendered configuration to each node:
+
[source,terminal]
----
$ oc describe machineconfigpool/worker
----
+

> Once the password resets after a node reboot, the machine config can be removed.

Reviewer 2: ```suggestion
Once the password is set after a node reboot, the machine config can be removed.
```

> Once the password resets after a node reboot, the machine config can be removed.

Reviewer 2: Maybe we should add something like:
You can verify that the password is set by trying to boot a roolback entry in the GRUB boot menu or y looking at the content of the `/boot/grub2/user.cfg` file.

Author: i've added a verification step

> Once the password resets after a node reboot, the machine config can be removed.

Reviewer: This line is a bit confusing. Do we want to say that applied MachineConfig can/should be removed once grub password has been set? In that case, newly scaled-up node won't have those grub config applied.

Reviewer 2: Oh, that's a good point. We need to keep it in.
